File: implementations/typescript/src/adapters/catalog.ts
import type {
  NativeImportStrategy,
  PackageManagerId,
  SupportTier,
} from "../domain/models.ts";

export interface PackageManagerDefinition {
  id: PackageManagerId;
  displayName: string;
  tier: SupportTier;
  aliases: string[];
  lockfiles: string[];
  configurationFiles: string[];
  installCommand: string[];
  implementationStatus: "production" | "preview";
  packageManagerPin: string;
  scope: string;
}

export const PACKAGE_MANAGERS: readonly PackageManagerDefinition[] = [
  {
    id: "npm",
    displayName: "npm",
    tier: "production-target",
    aliases: ["npm"],
    lockfiles: ["npm-shrinkwrap.json", "package-lock.json"],
    configurationFiles: [".npmrc"],
    installCommand: ["npm", "install"],
    implementationStatus: "production",
    packageManagerPin: "npm@12.0.2",
    scope: "npm manifests, lockfiles, workspaces, overrides, scripts, registries, CI, and containers",
  },
  {
    id: "pnpm",
    displayName: "pnpm",
    tier: "production-target",
    aliases: ["pnpm"],
    lockfiles: ["pnpm-lock.yaml"],
    configurationFiles: ["pnpm-workspace.yaml", ".npmrc", ".pnpmfile.cjs"],
    installCommand: ["pnpm", "install"],
    implementationStatus: "production",
    packageManagerPin: "pnpm@11.21.0",
    scope: "pnpm workspaces, catalogs, overrides, patches, linker policy, registries, CI, and containers",
  },
  {
    id: "yarn-classic",
    displayName: "Yarn Classic",
    tier: "production-target",
    aliases: ["yarn-classic", "yarn@1"],
    lockfiles: ["yarn.lock"],
    configurationFiles: [".yarnrc", ".npmrc"],
    installCommand: ["yarn", "install"],
    implementationStatus: "production",
    packageManagerPin: "yarn@1.22.22",
    scope: "Yarn 1 lockfiles, workspaces, resolutions, scripts, registries, CI, and containers",
  },
  {
    id: "yarn-modern",
    displayName: "Yarn Modern",
    tier: "production-target",
    aliases: ["yarn-modern", "yarn-berry"],
    lockfiles: ["yarn.lock"],
    configurationFiles: [".yarnrc.yml", ".pnp.cjs", ".pnp.loader.mjs"],
    installCommand: ["yarn", "install"],
    implementationStatus: "production",
    packageManagerPin: "yarn@4.18.0",
    scope: "Yarn modern workspaces, protocols, constraints, patches, plugins, linker modes, CI, and containers",
  },
  {
    id: "bun",
    displayName: "Bun",
    tier: "production-target",
    aliases: ["bun"],
    lockfiles: ["bun.lock", "bun.lockb"],
    configurationFiles: ["bunfig.toml", ".npmrc"],
    installCommand: ["bun", "install"],
    implementationStatus: "production",
    packageManagerPin: "bun@1.3.14",
    scope: "Bun lockfiles, workspaces, overrides, catalogs, scripts, registries, CI, and containers",
  },
  {
    id: "vlt",
    displayName: "vlt",
    tier: "production-target",
    aliases: ["vlt"],
    lockfiles: ["vlt-lock.json"],
    configurationFiles: ["vlt.json"],
    installCommand: ["vlt", "install"],
    implementationStatus: "production",
    packageManagerPin: "vlt@1.0.2",
    scope: "vlt manifests, lockfiles, workspaces, catalogs, graph modifiers, registries, CI, and containers",
  },
  {
    id: "deno",
    displayName: "Deno dependency mode",
    tier: "production-target",
    aliases: ["deno"],
    lockfiles: ["deno.lock"],
    configurationFiles: ["deno.json", "deno.jsonc", ".npmrc"],
    installCommand: ["deno", "install"],
    implementationStatus: "production",
    packageManagerPin: "deno@2.9.5",
    scope: "Deno npm-compatible dependency declarations, lockfiles, workspaces, registries, CI, and containers; runtime modernization remains out of scope",
  },
] as const;

export function getPackageManager(
  id: PackageManagerId,
): PackageManagerDefinition {
  const definition = PACKAGE_MANAGERS.find((candidate) => candidate.id === id);
  if (!definition) {
    throw new Error(`Unknown package manager definition: ${id}`);
  }
  return definition;
}

export function normalizePackageManagerId(
  value: string,
): PackageManagerId | null {
  const normalized = value.trim().toLowerCase();
  const exact = PACKAGE_MANAGERS.find((candidate) => candidate.id === normalized);
  if (exact) {
    return exact.id;
  }
  if (normalized === "yarn@1" || normalized === "yarn-1") {
    return "yarn-classic";
  }
  if (normalized === "yarn@modern" || normalized === "yarn-berry") {
    return "yarn-modern";
  }
  return null;
}

export function nativeImportStrategy(
  source: PackageManagerId,
  target: PackageManagerId,
  sourceLockfilePresent: boolean,
): NativeImportStrategy | null {
  if (!sourceLockfilePresent) return null;
  if (
    target === "pnpm"
    && ["npm", "yarn-classic", "yarn-modern"].includes(source)
  ) {
    return {
      id: "pnpm-import",
      source,
      target,
      mode: "dedicated-command",
      command: ["pnpm", "import"],
      summary: "Generate pnpm dependency state with pnpm's native lockfile importer.",
    };
  }
  if (
    target === "bun"
    && ["npm", "yarn-classic", "yarn-modern"].includes(source)
  ) {
    return {
      id: "bun-pm-migrate",
      source,
      target,
      mode: "dedicated-command",
      command: ["bun", "pm", "migrate"],
      summary: "Generate Bun dependency state with bun pm migrate.",
    };
  }
  if (source === "pnpm" && target === "bun") {
    return {
      id: "bun-pnpm-install-migration",
      source,
      target,
      mode: "install-integrated",
      command: installCommandFor(target),
      summary: "Use Bun's install-integrated pnpm lockfile migration path.",
    };
  }
  if (source === "npm" && target === "yarn-classic") {
    return {
      id: "yarn-classic-import",
      source,
      target,
      mode: "dedicated-command",
      command: ["yarn", "import"],
      summary: "Generate Yarn Classic dependency state with yarn import.",
    };
  }
  if (source === "yarn-classic" && target === "yarn-modern") {
    return {
      id: "yarn-modern-install-migration",
      source,
      target,
      mode: "install-integrated",
      command: installCommandFor(target),
      summary: "Use Yarn Modern's install-integrated Yarn Classic migration path.",
    };
  }
  if (source === "yarn-classic" && target === "npm") {
    return {
      id: "npm-yarn-lock-install",
      source,
      target,
      mode: "install-integrated",
      command: installCommandFor(target),
      summary: "Use npm's yarn.lock-aware installation path.",
    };
  }
  if (
    target === "deno"
    && source === "npm"
  ) {
    return {
      id: "deno-install-migration",
      source,
      target,
      mode: "install-integrated",
      command: installCommandFor(target),
      summary: "Use Deno's install-integrated Node dependency migration path.",
    };
  }
  return null;
}

function installCommandFor(target: PackageManagerId): string[] {
  const base = [...getPackageManager(target).installCommand];
  if (["npm", "pnpm", "yarn-classic", "bun"].includes(target)) {
    return [...base, "--ignore-scripts"];
  }
  if (target === "yarn-modern") return [...base, "--mode=skip-build"];
  return base;
}

File: implementations/rust/pkgshift-core/src/catalog.rs
use std::str::FromStr;

use crate::model::{
    ExecutableRequirement, NativeImportMode, NativeImportStrategy, PackageManagerId, SupportTier,
};

#[derive(Debug, Clone, Copy)]
pub struct PackageManagerDefinition {
    pub id: PackageManagerId,
    pub display_name: &'static str,
    pub tier: SupportTier,
    pub aliases: &'static [&'static str],
    pub lockfiles: &'static [&'static str],
    pub configuration_files: &'static [&'static str],
    pub install_command: &'static [&'static str],
    pub package_manager_pin: &'static str,
}

pub const PACKAGE_MANAGERS: &[PackageManagerDefinition] = &[
    PackageManagerDefinition {
        id: PackageManagerId::Npm,
        display_name: "npm",
        tier: SupportTier::ProductionTarget,
        aliases: &["npm"],
        lockfiles: &["npm-shrinkwrap.json", "package-lock.json"],
        configuration_files: &[".npmrc"],
        install_command: &["npm", "install", "--ignore-scripts"],
        package_manager_pin: "npm@12.0.2",
    },
    PackageManagerDefinition {
        id: PackageManagerId::Pnpm,
        display_name: "pnpm",
        tier: SupportTier::ProductionTarget,
        aliases: &["pnpm"],
        lockfiles: &["pnpm-lock.yaml"],
        configuration_files: &["pnpm-workspace.yaml", ".npmrc", ".pnpmfile.cjs"],
        install_command: &["pnpm", "install", "--ignore-scripts"],
        package_manager_pin: "pnpm@11.21.0",
    },
    PackageManagerDefinition {
        id: PackageManagerId::YarnClassic,
        display_name: "Yarn Classic",
        tier: SupportTier::ProductionTarget,
        aliases: &["yarn-classic", "yarn@1"],
        lockfiles: &["yarn.lock"],
        configuration_files: &[".yarnrc", ".npmrc"],
        install_command: &["yarn", "install", "--ignore-scripts"],
        package_manager_pin: "yarn@1.22.22",
    },
    PackageManagerDefinition {
        id: PackageManagerId::YarnModern,
        display_name: "Yarn Modern",
        tier: SupportTier::ProductionTarget,
        aliases: &["yarn-modern", "yarn-berry"],
        lockfiles: &["yarn.lock"],
        configuration_files: &[".yarnrc.yml", ".pnp.cjs", ".pnp.loader.mjs"],
        install_command: &["yarn", "install", "--mode=skip-build"],
        package_manager_pin: "yarn@4.18.0",
    },
    PackageManagerDefinition {
        id: PackageManagerId::Bun,
        display_name: "Bun",
        tier: SupportTier::ProductionTarget,
        aliases: &["bun"],
        lockfiles: &["bun.lock", "bun.lockb"],
        configuration_files: &["bunfig.toml", ".npmrc"],
        install_command: &["bun", "install", "--ignore-scripts"],
        package_manager_pin: "bun@1.3.14",
    },
    PackageManagerDefinition {
        id: PackageManagerId::Vlt,
        display_name: "vlt",
        tier: SupportTier::ProductionTarget,
        aliases: &["vlt"],
        lockfiles: &["vlt-lock.json"],
        configuration_files: &["vlt.json"],
        install_command: &["vlt", "install"],
        package_manager_pin: "vlt@1.0.2",
    },
    PackageManagerDefinition {
        id: PackageManagerId::Deno,
        display_name: "Deno dependency mode",
        tier: SupportTier::ProductionTarget,
        aliases: &["deno"],
        lockfiles: &["deno.lock"],
        configuration_files: &["deno.json", "deno.jsonc", ".npmrc"],
        install_command: &["deno", "install"],
        package_manager_pin: "deno@2.9.5",
    },
];

pub fn get_package_manager(id: PackageManagerId) -> &'static PackageManagerDefinition {
    PACKAGE_MANAGERS
        .iter()
        .find(|definition| definition.id == id)
        .expect("every package manager identifier has a catalog entry")
}

pub fn executable_requirement(id: PackageManagerId) -> ExecutableRequirement {
    let definition = get_package_manager(id);
    let program = definition
        .install_command
        .first()
        .expect("every package manager install command has a program");
    let (_, required_version) = definition
        .package_manager_pin
        .rsplit_once('@')
        .expect("every package manager pin includes an exact version");
    ExecutableRequirement {
        program: (*program).to_owned(),
        required_version: required_version.to_owned(),
        version_command: vec![(*program).to_owned(), "--version".to_owned()],
        package_manager_pin: definition.package_manager_pin.to_owned(),
    }
}

pub fn normalize_package_manager_id(value: &str) -> Option<PackageManagerId> {
    PackageManagerId::from_str(value).ok()
}

pub fn native_import_strategy(
    source: PackageManagerId,
    target: PackageManagerId,
    source_lockfile_present: bool,
) -> Option<NativeImportStrategy> {
    if !source_lockfile_present {
        return None;
    }
    let (id, mode, command, summary): (&str, NativeImportMode, &[&str], &str) =
        match (source, target) {
            (
                PackageManagerId::Npm
                | PackageManagerId::YarnClassic
                | PackageManagerId::YarnModern,
                PackageManagerId::Pnpm,
            ) => (
                "pnpm-import",
                NativeImportMode::DedicatedCommand,
                &["pnpm", "import"],
                "Generate pnpm dependency state with pnpm's native lockfile importer.",
            ),
            (
                PackageManagerId::Npm
                | PackageManagerId::YarnClassic
                | PackageManagerId::YarnModern,
                PackageManagerId::Bun,
            ) => (
                "bun-pm-migrate",
                NativeImportMode::DedicatedCommand,
                &["bun", "pm", "migrate"],
                "Generate Bun dependency state with bun pm migrate.",
            ),
            (PackageManagerId::Pnpm, PackageManagerId::Bun) => (
                "bun-pnpm-install-migration",
                NativeImportMode::InstallIntegrated,
                get_package_manager(target).install_command,
                "Use Bun's install-integrated pnpm lockfile migration path.",
            ),
            (PackageManagerId::Npm, PackageManagerId::YarnClassic) => (
                "yarn-classic-import",
                NativeImportMode::DedicatedCommand,
                &["yarn", "import"],
                "Generate Yarn Classic dependency state with yarn import.",
            ),
            (PackageManagerId::YarnClassic, PackageManagerId::YarnModern) => (
                "yarn-modern-install-migration",
                NativeImportMode::InstallIntegrated,
                get_package_manager(target).install_command,
                "Use Yarn Modern's install-integrated Yarn Classic migration path.",
            ),
            (PackageManagerId::YarnClassic, PackageManagerId::Npm) => (
                "npm-yarn-lock-install",
                NativeImportMode::InstallIntegrated,
                get_package_manager(target).install_command,
                "Use npm's yarn.lock-aware installation path.",
            ),
            (PackageManagerId::Npm, PackageManagerId::Deno) => (
                "deno-install-migration",
                NativeImportMode::InstallIntegrated,
                get_package_manager(target).install_command,
                "Use Deno's install-integrated Node dependency migration path.",
            ),
            _ => return None,
        };
    Some(NativeImportStrategy {
        id: id.to_owned(),
        source,
        target,
        mode,
        command: command.iter().map(ToString::to_string).collect(),
        summary: summary.to_owned(),
    })
}
